### DISCUSSION TITLE
What is the status of "CREATE FUNCTION"?

### DISCUSSION THREAD

**Discussion by qrilka** [apache/datafusion#12848] [Q&A] [answered]
I was doing some basic calculations using `datafusion-cli` and found that I repeat a couple of function invocations there. So I thought that it could be cool to have that in a function so I tried and saw this:

```
> create function foo();
Invalid or Unsupported Configuration: Function factory has not been configured``
```

Searching the documentation of Datafusion gave me no good results so I wonder what does this error mean (maybe it could be improved?) and if it's even planned to support UDFs defined from CLI and not in custom Rust code (which would need some extra work to plug it into datafusion-cli)

**Comment by Omega359** [comment, marked as answer]
The core framework to support it exists however it is otherwise unimplemented to the best of my knowledge. See 

- https://github.com/apache/datafusion/blob/main/datafusion/core/tests/user_defined/user_defined_scalar_functions.rs
- https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/function_factory.rs

**Comment by qrilka** [comment]
Thanks @Omega359 , but all of that seem to assume some custom Rust code, maybe `datafusion-cli` could report some more meaningful error like "custom functions are not supported in datafusion-cli"?

**Reply by findepi** [reply]
Better yet, could we have (perhaps in-memory) storage for functions within datafusion-cli?

**Reply by qrilka** [reply]
is it possible without custom Rust code for those functions?

**Reply by findepi** [reply]
If we cannot exemplify usage of create function APIs within DataFusion, how are users supposed to implement it?

**Reply by qrilka** [reply]
I guess that was the main part of my original question :)

**Reply by adriangb** [reply]
I think it would be helpful to have in-memory storage of functions not just in Datafusion CLI but also in other systems that use datafusion e.g. to create an ephemeral SQL function.

**Reply by alamb** [reply]
Did anyone make an issue / ticket to improve the error message? I agree that the current message is pretty obtuse

**Reply by qrilka** [reply]
@alamb I don't know about anything like that, the original question was about finding out the current situation/plans around UDFs i.e. about finding out whether such ticket is a valid one

**Reply by alamb** [reply]
Got it it -- thank you. If the error message is clear about the status then I agree nothing else is needed.

I was imagining we could help future users avoid confusion with a clearer error message

**Reply by qrilka** [reply]
I wouldn't say that the message is clear - if it were then this thread won't be needed. And I still don't understand the state of UDFs in `datafusion-cli` - I see that UDFs could be added by adding custom code but from a point of view of `datafusion-cli` user that doesn't look like an easy option: as I see it ATM to add a couple of simple function I'd need to create a custom fork of datafusion-cli with a small portion of extra Rust code.

**Reply by alamb** [reply]
>  I see that UDFs could be added by adding custom code but from a point of view of datafusion-cli user that doesn't look like an easy option: as I see it ATM to add a couple of simple function I'd need to create a custom fork of datafusion-cli with a small portion of extra Rust code.

Yes I agree -- user defined functions are not supported by datafusion-cli (or anything in the core crate either)
